redwood-protocol/src/commonMain/kotlin/app/cash/redwood/protocol/protocol.kt

```diff
@@ -32,16 +39,69 @@ import kotlinx.serialization.json.buildJsonArray
 import kotlinx.serialization.json.jsonArray
 import kotlinx.serialization.json.jsonPrimitive
 
-@Serializable
+@OptIn(ArrayContentSupport::class)
+@Serializable(EventSerializer::class)
 @Poko
 public class Event(
   /** Identifier for the widget from which this event originated. */
   public val id: Id,
   /** Identifies which event occurred on the widget with [id]. */
   public val tag: EventTag,
-  public val args: List<JsonElement> = emptyList(),
+  @ArrayContentBased public val args: Array<JsonElement> = emptyArray(),
 )
 
+private object EventSerializer : KSerializer<Event> {
```

Comment:
Hmm yeah this is unfortunate. I initially envisioned that we would be able to do this with a property-local serializer. But I guess that's too late because the enclosing serializer for the type is what determines whether or not the value needs to be included. Once we're inside a serializer for the property we are always going to write out the value.

I'll look into this more. Maybe ask the serialization folks if they'd be open to a similar annotation to treat arrays as values and not references.

Another potential option in the future is to use a `value class` when custom `equals` gets supported. Then we could have a `ValueBasedArray` wrapper around an array. This assumes that it doesn't somehow incur boxing, which is not yet clear since there's no implementation of this to play with yet.

Reply:
> I initially envisioned that we would be able to do this with a property-local serializer. But I guess that's too late because the enclosing serializer for the type is what determines whether or not the value needs to be included.

Yep, that appears to be the case.

I filed Kotlin/kotlinx.serialization#2409, let's see what they think.